documents/write_acls.go

```diff
@@ -84,3 +89,121 @@ func newChangedField(p proofs.Property, leaf *proofs.LeafNode, old bool) changed
 	cf.new = v
 	return cf
 }
+
+// transitionRulesFor returns a copy all the transition rules for the account.
+func (cd *CoreDocument) transitionRulesFor(account identity.DID) (rules []coredocumentpb.TransitionRule) {
+	for _, rule := range cd.Document.TransitionRules {
+		for _, rk := range rule.Roles {
+			role, err := getRole(rk, cd.Document.Roles)
+			if err != nil {
+				continue
+			}
+
+			if _, ok := isAccountInRole(role, account); !ok {
+				continue
+			}
+
+			rules = append(rules, coredocumentpb.TransitionRule{
+				RuleKey:   copyBytes(rule.RuleKey),
+				Roles:     copyByteSlice(rule.Roles),
+				MatchType: rule.MatchType,
+				Field:     copyBytes(rule.Field),
+				Action:    rule.Action,
+			})
+		}
+	}
+
+	return rules
+}
+
+func copyBytes(data []byte) []byte {
+	if data == nil {
+		return nil
+	}
+
+	nb := make([]byte, len(data), len(data))
+	copy(nb, data)
+	return nb
+}
+
+func copyByteSlice(data [][]byte) [][]byte {
+	nbs := make([][]byte, len(data), len(data))
+	for i, b := range data {
+		nbs[i] = copyBytes(b)
+	}
+
+	return nbs
+}
+
+// validateTransitions validates the changedFields based on the rules provided.
+// returns an error if any changedField violates the rules.
+func validateTransitions(rules []coredocumentpb.TransitionRule, changedFields []changedField) error {
+	cfMap := make(map[string]struct{})
+	for _, cf := range changedFields {
+		cfMap[cf.name] = struct{}{}
+	}
+
+	for _, rule := range rules {
+		for _, cf := range changedFields {
+			if isValidTransition(rule, cf) {
+				delete(cfMap, cf.name)
```

Comment:
Why don't you just append the error here, if its not a valid transition?

Reply:
It may be an invalid transition for this rule doesn't mean its invalid altogether. We can only be sure about the transition is invalid once we iterate all the changes to all the rules